src/test/java/com/ita/if103java/ims/service/WarehouseServiceImplTest.java

```diff
@@ -0,0 +1,454 @@
+        assertEquals(warehouseService.findAllTopLevel(pageable, userDetails), page);
+    }
+
+    @Test
+    void update_success(){
+        Warehouse topLevel = new Warehouse(4L, "updatedWarehouse", "auto parts", null, false, null, 1L, 4L, true);
+
+        Warehouse updatedWarehouse = new Warehouse(12L, "updatedWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        Warehouse dBWarehouse = new Warehouse(12L, "dBWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        warehouseDto = new WarehouseDto(12L, "WarehouseTest", "auto parts", 20, true, 5L, 1L, 4L, true, addressDto);
+
+        when(warehouseDtoMapper.toEntity(warehouseDto)).thenReturn(updatedWarehouse);
+        when(warehouseDao.findById(updatedWarehouse.getId(), dBWarehouse.getAccountID())).thenReturn(dBWarehouse);
+        assertTrue(dBWarehouse.isActive());
+        when(addressDtoMapper.toEntity(warehouseDto.getAddressDto())).thenReturn(address);
+        Event event = new Event("Warehouse edited " +
+            updatedWarehouse.getName(), updatedWarehouse.getAccountID(),
+            updatedWarehouse.getId(), 1L, EventName.WAREHOUSE_EDITED, 2L);
+        doNothing().when(eventService).create(event);
+        when(warehouseDao.update(dBWarehouse)).thenReturn(updatedWarehouse);
+        assertFalse(updatedWarehouse.isTopLevel());
+
+        when(warehouseDao.findLevelByParentID(dBWarehouse.getParentID())).thenReturn(2);
+      when(warehouseDao.findByTopWarehouseID(4L, 1L)).thenReturn(List.of(new Warehouse()));
+        assertEquals(warehouseService.update(warehouseDto, userDetails), updatedWarehouse);
+    }
+
+    @Test
+    void update_notActive() {
+        Warehouse inActive = new Warehouse(12L, "WarehouseTest", "auto parts", 20, true, 5L, 1L, 4L, false);
+        when(warehouseDtoMapper.toEntity(warehouseDto)).thenReturn(inActive);
+        assertFalse(warehouse.isActive());
+        WarehouseUpdateException exception = assertThrows(WarehouseUpdateException.class, () -> {
+            warehouseService.update(warehouseDto, userDetails);
+        });
+        assertEquals("You can't make warehouse inactive!", exception.getMessage());
+    }
+
+    @Test
+    void update_notAllowChangeParent() {
+        Warehouse updatedWarehouse = new Warehouse(12L, "updatedWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        Warehouse dBWarehouse = new Warehouse(12L, "dBWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        warehouseDto = new WarehouseDto(12L, "WarehouseTest", "auto parts", 20, true, 5L, 1L, 4L, true, null);
+
+        when(warehouseDtoMapper.toEntity(warehouseDto)).thenReturn(updatedWarehouse);
+        when(warehouseDao.findById(updatedWarehouse.getId(), updatedWarehouse.getAccountID())).thenReturn(dBWarehouse);
+        warehouseDto.setId(3L);
+        assertTrue(updatedWarehouse.isActive());
+        updatedWarehouse.setActive(true);
+        if (!updatedWarehouse.getParentID().equals(dBWarehouse.getParentID())) {
```

Comment:
there are should not be an if condition in the test.
Your test may never execute code inside if and then it will test nothing.
Please remove this **if** condition and update your test that this condition will always be true